gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientLibraryWriteChannel.java

```diff
@@ -0,0 +1,209 @@
+  private WriteChannel writeChannel;
+  private final CreateObjectOptions createOptions;
+  private Boolean uploadSuccess = false;
+  // TODO: not supported as of now
+  // private final String requesterPaysProject;
+
+  public GoogleCloudStorageClientLibraryWriteChannel(
+      Storage storage,
+      GoogleCloudStorageOptions storageOptions,
+      StorageResourceId resourceId,
+      CreateObjectOptions createOptions,
+      ExecutorService uploadThreadPool) {
+    super(uploadThreadPool, storageOptions.getWriteChannelOptions());
+    this.storage = storage;
+    this.storageOptions = storageOptions;
+    this.resourceId = resourceId;
+    this.createOptions = createOptions;
+    this.channelOptions = storageOptions.getWriteChannelOptions();
+    this.writeChannel =
+        getClientLibraryWriteChannel(storage, resourceId, createOptions, storageOptions);
+  }
+
+  @Override
+  public void startUpload(InputStream pipeSource) throws IOException {
+    // Given that the two ends of the pipe must operate asynchronous relative
+    // to each other, we need to start the upload operation on a separate thread.
+    try {
+      uploadOperation = threadPool.submit(new UploadOperation(pipeSource, this.resourceId));
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Failed to start upload for '%s'", resourceId), e);
+    }
+  }
+
+  private BlobInfo getBlobInfo(StorageResourceId resourceId, CreateObjectOptions createOptions) {
+    BlobId blobId =
+        BlobId.of(
+            resourceId.getBucketName(), resourceId.getObjectName(), resourceId.getGenerationId());
+    BlobInfo blobInfo =
+        BlobInfo.newBuilder(blobId)
+            .setContentType(createOptions.getContentType())
+            .setContentEncoding(createOptions.getContentEncoding())
+            .setMetadata(encodeMetadata(createOptions.getMetadata()))
+            .build();
+    return blobInfo;
+  }
+
+  private WriteChannel getClientLibraryWriteChannel(
+      Storage storage,
+      StorageResourceId resourceId,
+      CreateObjectOptions createOptions,
+      GoogleCloudStorageOptions storageOptions) {
+    WriteChannel writeChannel =
+        storage.writer(getBlobInfo(resourceId, createOptions), generateWriteOptions());
+    writeChannel.setChunkSize(
+        Math.toIntExact(
+            storageOptions.getWriteChannelOptions().getNumberOfBufferedRequests()
+                * MAX_WRITE_CHUNK_BYTES.getNumber()));
+
+    return writeChannel;
+  }
+
+  private class UploadOperation implements Callable<Boolean> {
+
+    // Read end of the pipe.
+    private final BufferedInputStream pipeSource;
+    private final StorageResourceId resourceId;
+    private final int MAX_BYTES_PER_MESSAGE = MAX_WRITE_CHUNK_BYTES.getNumber();
+
+    UploadOperation(InputStream pipeSource, StorageResourceId resourceId) {
+      this.resourceId = resourceId;
+      this.pipeSource = new BufferedInputStream(pipeSource, MAX_BYTES_PER_MESSAGE);
```

Comment:
I think that have dedicated config option to configure pipe and buffer size.

Reply:
It's same configuration which is used in `GoogleCloudStorageGrpcWriteChannel`.

Reply:
I think that `GoogleCloudStorageGrpcWriteChannel` wasn't completely productionized - we already support these configs for non-gRPC channel, so probably we should implement the same logic from get go here too.